Fix force reloading last transition from database

jacobpgn opened the pull request:
The [current implementation](https://github.com/gocardless/statesman/blob/14250521d38449244e9bcb8f5b31b76e26614c39/lib/statesman/adapters/active_record.rb#L44-L60) for force reloading will use the
in-memory transitions if they're available instead of forcing a DB query
like you'd expect when passing this flag.

This means that `force_reload: true` currently forces a cache-level
reload of the most recent transition, but there is no way to force a
DB-level reload when the transitions are in memory.

In most cases the current behaviour works as expected – accessing
`model.state_machine.current_state` will issue a SELECT with LIMIT 1,
so the full set of the model's transitions will not be loaded into
memory and subsequent calls to `current_state` will issue fresh queries.
This breaks when you actually load the association, e.g. with
`model.history.inspect`. Once this has happened, the adapter's `history`
(and thus also methods such as `model.state_machine.current_state`) will
use the in-memory transitions whether `force_reload` is true or false,
and never issue fresh queries.

In practice, you often want a way to reload the transitions from the DB
to ensure that you pick up on any new transitions from other processes,
or other instances of the same model in the same process. This is what
is expected when you explicitly ask statesman to `force_reload`!

This change simply passes `force_reload` along to `history`, and checks
that it's false before going ahead and using in-memory transitions.

hmac commented:
I think this is a good idea. With `force_reload` and the AR adapter you are invariably going to want to fetch the latest transition from the database rather from the AR cache. 

I reckon this warrants a minor version bump (because of the change in behaviour) and a clear explanation in the changelog.

petehamilton commented:
Bonus points for a great branch name.

jacobpgn commented:
Thanks team!

I've squashed and updated the commit message – chatting with @petehamilton earlier we realised the description of the cause wasn't quite right. It should be correct now, so I'd love it if you could do a quick pass on the wording and see if it makes sense to you.

I'm planning to roll this and https://github.com/gocardless/statesman/pull/285 out as a minor release (https://github.com/gocardless/statesman/pull/285 is a non-breaking feature, and this PR is a bugfix). Seem legit? cc @timrogers

timrogers commented:
I wonder if we can find a clearer title for the commit which specifies what exactly our "fix" is - how about something like "Force reload from the DB, not the cache, in the ActiveRecord adapter" (I tried to get this down to [50 characters](http://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html) but I couldn't do this while also making it declarative, and I'd rather have an informative message than an artificially short one).

petehamilton commented:
@timrogers If I can't make 50, I generally aim for `<= 72` which is where Github starts displaying things weirdly and is also less than the standard `80` character wrapping width for most people's setup.

timrogers commented:
Just realised my idea isn't a very good description of the substance of the change. Let me have a think.

petehamilton commented:
> Fix force_reload for active record models with preloaded transitions?

timrogers commented on lib/statesman/adapters/active_record.rb:
```diff
@@ -41,8 +41,8 @@ def create(from, to, metadata = {})
 
-      def history
-        if transitions_for_parent.loaded?
+      def history(force_reload: false)
```
Is this not slightly odd, in that it means that the interface of the ActiveRecord adapter differs from the other adapters? `#history` is a public part of the adapter interface, and the `Mongoid` and `Memory` adapters won't support this additional parameter. It might be good to expand this change to them too.

jacobpgn replied:
I don't think so: each of the adapters currently has a different interface! If you're explicitly calling the adapter's methods then you should know what interface you have available for that adapter.

timrogers replied:
I'm not convinced that's quite true - e.g. `Memory#last` seems to accept any argument, presumably for compatibility with other adapters:

https://github.com/gocardless/statesman/blob/b962ed219e6b05328694def8f9a1d5a504ec5c2e/lib/statesman/adapters/memory.rb#L31

timrogers commented on lib/statesman/adapters/active_record.rb:
```diff
@@ -53,7 +53,7 @@ def history
 
       def last(force_reload: false)
```
Do we want to make sure the other adapters don't have this problem? The memory adapter is safe:

https://github.com/gocardless/statesman/blob/b962ed219e6b05328694def8f9a1d5a504ec5c2e/lib/statesman/adapters/memory.rb#L31-L33

...but the Mongoid adapter has the same bug:

https://github.com/gocardless/statesman/blob/b962ed219e6b05328694def8f9a1d5a504ec5c2e/lib/statesman/adapters/mongoid.rb#L39-L45
